### PR TITLE
[Test] SFPU edge case testing phase 4, rebuilt on the current test infra

### PR DESCRIPTION
**Pull request by ldjurovicTT**
Re-implements PR #54727 on top of today's main, where the SFPU test infrastructure has since been simplified (#56109, #55934, #55955, #55958): reason-string dicts became frozensets with grouped comments, the binary suite gained the Dest/pack golden modelling and retracted its negative-zero xfails, and the ReluMin #1120 skip is gone. Same coverage, ~800 fewer added lines, and adapted rather than transcribed where main had already moved.

Drives the edge classes the suite was not reaching, and makes the goldens model the datapath so those probes can be asserted instead of xfailed.

Ternary family, cat B. test_sfpu_ternary_operand_edges replaces test_sfpu_ternary_edges with two runtime axes -- which operand carries the probe, and which failure class it belongs to (pole / nan_in / inf_in). The driver asserts a whole tile at once, so partitioning before driving is what keeps addcdiv's c = 0, c = NaN and c = inf out of one variant and one xfail. Five divergences remain, in three causes, each registered against the one class it was measured on. TernarySFPUGolden and WhereGolden now model the block-float unpack quantization, the store into a Dest whose width dest_acc selects, and the pack out of it -- without those three steps a probe reads the packer's substituted infinity as the kernel having computed one, the defect behind the binary suite's retracted "0/0 returns inf" xfails.

Unary family, cat F. New format_extremes() / extreme_values() and their two gates, a table-driven saturation sweep, and 69 ops enrolled against a measurement. subnormal_delivered() keeps the subnormal probe off the pipelines that flatten it, measured with the four ops whose answer there differs from their answer at zero.

Binary and integer. Integer zero as both operand and knee; the uint32 upper half, now reaching 2**32 - 1; signed division separating truncation from floor; logsigmoid's exp branch, the one that reads operand B, with the golden modelling it and the derived (x, exp(-x)) probe split so the NaN pair's sign divergence does not excuse the four pairs measured to agree; a saturation sweep whose pairs are asserted to straddle the ceiling.

Block-float. New unary and ternary variants whose 16-datum blocks deliberately span the shared exponent, at 4, 12 and 24 decades of spread.

Signed zero at a pole (cat G). boundary_probes() now emits -0.0 at a registered zero pole, gated on delivery alone. This class was 0 ops before and is 17 now. It also found RsqrtCompat(-0.0) = +inf, and moved Sqrt and Rsqrt off the cat-B derivation onto delivery, which is where their probe now comes from.

A coverage ratchet. Coverage loss here is silent -- drop an op from an enrolment table and the sweep collects fewer variants and stays green. coverage_counts() asks the stimulus machinery, not a hand-written ledger, and test_coverage_does_not_regress turns a regression into a red run.

Test only: no LLK, kernel or runtime source is touched. 167 host-only tests pass (128 before); the device suites collect 7389 / 2520 / 224 variants. The divergences carried over are measurements from the original PR's Blackhole p150 and Wormhole n300 runs, non-strict so a behaviour change reports XPASS.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ldjurovic/sfpu_edge_cases_phase_4_compact)

### CODE REVIEW

File: tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py [hunk 1]
```diff
@@ -4205,12 +4206,25 @@ def _isclose(self, t1, t2):
...
+
+        Of the kernel's three branches the golden models only x > 4, where it returns `-t2`:
+        the only branch whose result depends on operand B, so torch there cannot tell a kernel
+        that used the operand from one that ignored it -- and torch's -log1p(exp(-x)) differs
+        from -exp(-x) by up to 0.9% just above the threshold. test_sfpu_domains pins the
+        approximation separately, so modelling the branch does not stop asserting that
+        -exp(-x) is a logsigmoid.
+        """
+        x = t1.to(torch.float32)
+        if float(x) > self._LOGSIGMOID_EXP_BRANCH:
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py:4281]: Added, with the wording kept as the message: `assert x.numel() == 1, "per-element golden; vectorize the exp-branch test before batching rows"`.

Agreed on the diagnosis — the caller loop makes it scalar today, and the docstring read as though it were not. The docstring now says per element, and says why the assert is there rather than just what it checks, since the hazard is the refactor and not the current code.

`test_logsigmoid_golden_refuses_a_batched_row` pins it: a two-element row straddling the threshold must raise, and the scalars either side of the branch must still work.

File: tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py [hunk 2]
```diff
@@ -5304,7 +5406,32 @@ def __call__(
...
+        # goldens do at the same point. See the class comment for why it is unconditional.
+        emitted_nan = torch.isnan(result)
+        result = torch.where(emitted_nan, result.abs(), result)
+
+        # Two casts, both NaN-sign preserving: the Dest write's narrowing, then the store out
+        # to the output format, which is not always the Dest format. Plain .to() would use
+        # torch's canonicalising bf16 cast, forcing every NaN's sign bit to 1 and deciding the
+        # substituted infinity's sign by accident.
+        result = cast_to_dest_dtype(result, format_dict[dst_format]).float()
+        result = cast_to_dest_dtype(result, format_dict[data_format]).flatten()
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py:5483]: Correct, and it matters more than the tolerance made it look. The gap was pre-existing — `test_sfpu_ternary` has driven `Bfp8_b -> Bfp8_b` addcmul for a while — but `test_sfpu_ternary_block_spread` is the variant whose entire stimulus is a block spanning the shared exponent, so it was passing for the wrong reason: `passed_test()`'s `_bfp_block_aware_compare` fallback exists to excuse exactly the difference the golden was not modelling.

Fixed with a new `quantize_output_to_pack_format()`, the counterpart of `quantize_input_to_unpack_format()` and the same flat 16-datum grouping (the kernel is layout-preserving, so golden and device agree about where a block starts).

Two blocks it deliberately does not hand to the converter, both found while writing it:

* a block holding a non-finite has no shared exponent left — the non-finite takes it and every finite neighbour with it, which is `_bfp_zero_nonfinite_blocks`' job;
* an **all-zero** block makes `bfp8b_to_float16b()` decode the leading 1 a zero does not have and answer `2^-127` on every lane. That is a pre-existing quirk in the shared helper, reachable from the input side too; not fixed here, because every other suite's goldens go through it and the blast radius does not belong in this PR. Guarded against instead, so the new path cannot produce a golden the packer never writes.

Two host tests: `test_ternary_golden_quantizes_a_block_float_output` asserts the low lane of a wide block is actually gone (the one observable difference between the two goldens), and `test_pack_quantizer_leaves_a_block_it_cannot_encode_alone` covers the two skip cases per block, with a quantizable neighbour to show the skip is not tensor-wide.

File: tt_metal/tt-llk/tests/python_tests/helpers/sfpu_domains.py [hunk 3]
```diff
@@ -1657,6 +1677,70 @@ def format_specials(fmt: DataFormat) -> Tuple[float, ...]:
...
+    if fmt not in _FORMAT_MIN_NORMAL:
+        raise ValueError(
+            f"{fmt.name} has no per-element smallest normal (see _FORMAT_MIN_NORMAL): its "
+            "small end is set by the exponent shared across a block, so a cat-F probe would "
+            "be wrong for every block but one"
+        )
+    ceiling_fmt = fmt if fmt in _FORMAT_MAX_MAGNITUDE else DataFormat.Float16_b
+    ceiling = _truncate_mantissa(
+        _FORMAT_MAX_MAGNITUDE.get(fmt, _BF16_MAX_MAGNITUDE), ceiling_fmt
+    )
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tt_metal/tt-llk/tests/python_tests/helpers/sfpu_domains.py]: Right, and thanks — this was silent in the worst way: 3.3895e38 round-trips through fp32 intact, so `test_format_extremes_straddle_the_ftz_cliff_and_are_representable[Float32]` passed while the thirty-thousandth of a decade up to fp32's real ceiling went unprobed, and the small end was already fp32-specific.

The obvious fix — adding `Float32` to `_FORMAT_MAX_MAGNITUDE` — is wrong, and the suite caught it. `Float32` is absent there **on purpose**: the fallback makes it tie with `Float16_b`, `narrowest_range_format()` resolves that tie to its first argument (the stimulus format), and that is what spaces the acosh probe on a `Float32 -> Float16_b` pipeline by an fp32 ULP rather than a bfloat16 one. Adding the row turned `test_acosh_above_pole_probe_is_widened_at_dest_acc_no` red at `1.015625` against `1.0000002384185791`.

So the fix separates the two questions the table had conflated:

* `_FORMAT_MAX_MAGNITUDE` keeps its meaning, "what bounds a domain", and its comment now says why `Float32` stays out;
* new `format_max_magnitude()` answers "what can this format hold", and `format_extremes()` and `clip_to_format()` ask it — the clip mattered, or the probe would have been emitted and then dropped;
* `extreme_values()` bounds against a new `narrowest_ceiling_format()`, because a magnitude probe has to survive **both** legs. Without that, `Float32 -> Float16_b` started emitting 3.40e38 into a bfloat16 output, which made `Gelu` invent a NaN at a cat-F probe and tripped the `_EXTREMES_NAN_OPS <= SPECIALS_READY_OPS` gate — the pack's saturation is the saturation sweep's subject, not cat F's.

Both saturation host asserts now classify against fp32's real ceiling too, which is what the "stay clear of the band between bfloat16's ceiling and fp32's" comment above them always claimed. Pinned by `test_float32_ceiling_is_probed_as_its_own_and_not_bfloat16s` and `test_the_fp32_ceiling_probe_stays_off_a_narrower_output`, the latter asserting that the two "narrowest" helpers keep their different answers on that pair.
